File: src/python/xerxes/runtime/tool_pool.py
# Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Filtered, schema-ready view of an :class:`ExecutionRegistry` tool set.

:class:`ToolPool` is an immutable snapshot of the tools that should be
exposed to the LLM for one turn (or one sub-agent), already filtered by
category, deny-list, safety, and MCP-inclusion rules. :func:`assemble_tool_pool`
builds the pool from a populated registry.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from .execution_registry import ExecutionRegistry, RegistryEntry


@dataclass(frozen=True)
class ToolPool:
    """Immutable filtered set of tools exposed to one turn or sub-agent.

    Attributes:
        tools: Tuple of registry entries that survived filtering.
        denied_tools: Tool names that were explicitly excluded.
        categories: Categories the pool was filtered to (empty = all).
        safe_only: Whether only safe (no-permission-prompt) tools were kept.
    """

    tools: tuple[RegistryEntry, ...] = ()
    denied_tools: frozenset[str] = frozenset()
    categories: tuple[str, ...] = ()
    safe_only: bool = False

    @property
    def tool_count(self) -> int:
        """Number of tools surviving the filter chain."""
        return len(self.tools)

    @property
    def tool_names(self) -> tuple[str, ...]:
        """Tuple of tool names in the pool, in registry order."""
        return tuple(t.name for t in self.tools)

    def get_tool(self, name: str) -> RegistryEntry | None:
        """Look up an entry by exact name; ``None`` if not in the pool."""
        for t in self.tools:
            if t.name == name:
                return t
        return None

    def to_schemas(self) -> list[dict[str, Any]]:
        """Return JSON tool schemas suitable for sending to the LLM."""

        schemas = []
        for entry in self.tools:
            if entry.schema:
                schemas.append(entry.schema)
            else:
                schemas.append(
                    {
                        "name": entry.name,
                        "description": entry.description or f"Execute {entry.name}",
                        "input_schema": {"type": "object", "properties": {}},
                    }
                )
        return schemas

    def as_markdown(self) -> str:
        """Render the pool as a Markdown overview (counts, filters, tool list)."""

        lines = [
            "# Tool Pool",
            "",
            f"Tools: {self.tool_count}",
            f"Safe only: {self.safe_only}",
            f"Categories: {', '.join(self.categories) or 'all'}",
            f"Denied: {', '.join(sorted(self.denied_tools)) or 'none'}",
            "",
        ]
        for tool in self.tools:
            safe_tag = " [safe]" if tool.safe else ""
            cat_tag = f" ({tool.category})" if tool.category else ""
            lines.append(f"- **{tool.name}**{safe_tag}{cat_tag} — {tool.description}")
        return "\n".join(lines)


def assemble_tool_pool(
    registry: ExecutionRegistry | None = None,
    categories: list[str] | None = None,
    deny_tools: set[str] | None = None,
    deny_prefixes: list[str] | None = None,
    safe_only: bool = False,
    include_mcp: bool = True,
) -> ToolPool:
    """Build a filtered :class:`ToolPool` from ``registry``.

    Args:
        registry: Populated :class:`ExecutionRegistry`; an empty pool is
            returned when ``None``.
        categories: When provided, restrict tools to these categories.
        deny_tools: Set of tool names to exclude entirely.
        deny_prefixes: Tools whose name starts with any of these prefixes
            are excluded (useful for blocking ``mcp__foo__bar`` etc.).
        safe_only: When ``True`` keep only tools flagged ``safe``.
        include_mcp: When ``False`` drop tools whose ``source_hint`` mentions
            ``"mcp"``.
    """

    if registry is None:
        return ToolPool()

    deny = deny_tools or set()
    prefixes = deny_prefixes or []

    tools = registry.list_tools(safe_only=safe_only)

    if categories:
        tools = [t for t in tools if t.category in categories]

    if deny:
        tools = [t for t in tools if t.name not in deny]

    if prefixes:
        tools = [t for t in tools if not any(t.name.startswith(p) for p in prefixes)]

    if not include_mcp:
        tools = [t for t in tools if "mcp" not in t.source_hint.lower()]

    return ToolPool(
        tools=tuple(tools),
        denied_tools=frozenset(deny),
        categories=tuple(categories or []),
        safe_only=safe_only,
    )


__all__ = [
    "ToolPool",
    "assemble_tool_pool",
]

File: xerxes/src/runtime/toolPool.ts
// Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
// Licensed under the Apache License, Version 2.0.

import type { ExecutionRegistry, RegistryEntry } from './executionRegistry.js'

export interface ToolPoolOptions {
  readonly categories?: readonly string[]
  readonly deniedTools?: Iterable<string>
  readonly safeOnly?: boolean
  readonly tools?: readonly RegistryEntry[]
}

export interface AssembleToolPoolOptions {
  /** Restrict the pool to matching tool categories. An empty list keeps all categories. */
  readonly categories?: readonly string[]
  /** Tool names to exclude exactly. */
  readonly denyTools?: Iterable<string>
  /** Exclude tool names beginning with any supplied prefix. */
  readonly denyPrefixes?: readonly string[]
  /** Keep only tools which the registry marks safe. */
  readonly safeOnly?: boolean
  /** Omit tools whose source hint contains mcp, case-insensitively. */
  readonly includeMcp?: boolean
}

/**
 * Immutable schema-ready view of tools exposed for one turn or sub-agent.
 *
 * Entries are snapshotted at construction time, so later registry mutations
 * and caller-owned filter collection mutations do not alter this pool.
 */
export class ToolPool {
  readonly categories: readonly string[]
  readonly deniedTools: ReadonlySet<string>
  readonly safeOnly: boolean
  readonly tools: readonly RegistryEntry[]

  constructor(options: ToolPoolOptions = {}) {
    this.tools = Object.freeze((options.tools ?? []).map(snapshotEntry))
    this.deniedTools = immutableSet(options.deniedTools ?? [])
    this.categories = Object.freeze([...(options.categories ?? [])])
    this.safeOnly = options.safeOnly ?? false
    Object.freeze(this)
  }

  get toolCount(): number {
    return this.tools.length
  }

  get toolNames(): readonly string[] {
    return Object.freeze(this.tools.map(tool => tool.name))
  }

  /** Look up a tool by its exact advertised name. */
  getTool(name: string): RegistryEntry | undefined {
    return this.tools.find(tool => tool.name === name)
  }

  /** Return immutable tool schemas suitable for an LLM request. */
  toSchemas(): ReadonlyArray<Readonly<Record<string, unknown>>> {
    return Object.freeze(this.tools.map(toolSchema))
  }

  /** Render the pool's filters and surviving tools as a Markdown overview. */
  asMarkdown(): string {
    const lines = [
      '# Tool Pool',
      '',
      'Tools: ' + this.toolCount,
      'Safe only: ' + this.safeOnly,
      'Categories: ' + (this.categories.join(', ') || 'all'),
      'Denied: ' + ([...this.deniedTools].sort().join(', ') || 'none'),
      '',
    ]
    for (const tool of this.tools) {
      const safeTag = tool.safe ? ' [safe]' : ''
      const categoryTag = tool.category ? ' (' + tool.category + ')' : ''
      lines.push('- **' + tool.name + '**' + safeTag + categoryTag + ' — ' + tool.description)
    }
    return lines.join('\n')
  }
}

/**
 * Build an immutable filtered tool pool from a registry.
 *
 * Filtering follows the Python runtime's ordering: safety, categories, exact
 * deny-list names, denied prefixes, and then MCP source hints.
 */
export function assembleToolPool(
  registry: ExecutionRegistry | null | undefined = undefined,
  options: AssembleToolPoolOptions = {},
): ToolPool {
  if (registry === undefined || registry === null) return new ToolPool()

  const denied = new Set(options.denyTools ?? [])
  const prefixes = options.denyPrefixes ?? []
  const categories = options.categories ?? []
  const safeOnly = options.safeOnly ?? false
  const includeMcp = options.includeMcp ?? true

  let tools = registry.listTools({ safeOnly })
  if (categories.length > 0) tools = tools.filter(tool => categories.includes(tool.category))
  if (denied.size > 0) tools = tools.filter(tool => !denied.has(tool.name))
  if (prefixes.length > 0) tools = tools.filter(tool => !prefixes.some(prefix => tool.name.startsWith(prefix)))
  if (!includeMcp) tools = tools.filter(tool => !tool.sourceHint.toLowerCase().includes('mcp'))

  return new ToolPool({
    tools,
    deniedTools: denied,
    categories,
    safeOnly,
  })
}

function toolSchema(entry: RegistryEntry): Readonly<Record<string, unknown>> {
  if (hasSchema(entry.schema)) return entry.schema
  return Object.freeze({
    name: entry.name,
    description: entry.description || 'Execute ' + entry.name,
    input_schema: Object.freeze({
      type: 'object',
      properties: Object.freeze({}),
    }),
  })
}

function hasSchema(schema: Readonly<Record<string, unknown>> | undefined): schema is Readonly<Record<string, unknown>> {
  return schema !== undefined && Object.keys(schema).length > 0
}

function snapshotEntry(entry: RegistryEntry): RegistryEntry {
  return Object.freeze({
    name: entry.name,
    kind: entry.kind,
    description: entry.description,
    handler: entry.handler,
    category: entry.category,
    safe: entry.safe,
    sourceHint: entry.sourceHint,
    schema: entry.schema === undefined ? undefined : freezeRecord(entry.schema),
  })
}

function freezeRecord(record: Readonly<Record<string, unknown>>): Readonly<Record<string, unknown>> {
  const snapshot: Record<string, unknown> = {}
  for (const [key, value] of Object.entries(record)) snapshot[key] = freezeValue(value)
  return Object.freeze(snapshot)
}

function freezeValue(value: unknown): unknown {
  if (Array.isArray(value)) return Object.freeze(value.map(freezeValue))
  if (value instanceof Date) return value.toISOString()
  if (value !== null && typeof value === 'object') return freezeRecord(value as Readonly<Record<string, unknown>>)
  return value
}

function immutableSet(values: Iterable<string>): ReadonlySet<string> {
  const valuesSnapshot = new Set(values)
  const snapshot: ReadonlySet<string> = {
    get size(): number {
      return valuesSnapshot.size
    },
    has: value => valuesSnapshot.has(value),
    entries: () => valuesSnapshot.entries(),
    keys: () => valuesSnapshot.keys(),
    values: () => valuesSnapshot.values(),
    forEach: (callback, thisArg) => {
      valuesSnapshot.forEach(value => callback.call(thisArg, value, value, snapshot))
    },
    [Symbol.iterator]: () => valuesSnapshot[Symbol.iterator](),
    union: <Value>(other: ReadonlySetLike<Value>): Set<string | Value> => {
      const result = new Set<string | Value>(valuesSnapshot)
      forEachSetValue(other, value => result.add(value))
      return result
    },
    intersection: <Value>(other: ReadonlySetLike<Value>): Set<string & Value> => {
      const result = new Set<string & Value>()
      for (const value of valuesSnapshot) {
        if (other.has(value as unknown as Value)) result.add(value as unknown as string & Value)
      }
      return result
    },
    difference: <Value>(other: ReadonlySetLike<Value>): Set<string> => {
      const result = new Set<string>()
      for (const value of valuesSnapshot) {
        if (!other.has(value as unknown as Value)) result.add(value)
      }
      return result
    },
    symmetricDifference: <Value>(other: ReadonlySetLike<Value>): Set<string | Value> => {
      const result = new Set<string | Value>(valuesSnapshot)
      forEachSetValue(other, value => {
        if (valuesSnapshot.has(value as unknown as string)) result.delete(value)
        else result.add(value)
      })
      return result
    },
    isSubsetOf: (other: ReadonlySetLike<unknown>): boolean => {
      for (const value of valuesSnapshot) {
        if (!other.has(value)) return false
      }
      return true
    },
    isSupersetOf: (other: ReadonlySetLike<unknown>): boolean => {
      let result = true
      forEachSetValue(other, value => {
        if (!valuesSnapshot.has(value as string)) result = false
      })
      return result
    },
    isDisjointFrom: (other: ReadonlySetLike<unknown>): boolean => {
      let result = true
      forEachSetValue(other, value => {
        if (valuesSnapshot.has(value as string)) result = false
      })
      return result
    },
  }
  return Object.freeze(snapshot)
}

function forEachSetValue<Value>(set: ReadonlySetLike<Value>, callback: (value: Value) => void): void {
  const iterator = set.keys()
  for (;;) {
    const next = iterator.next()
    if (next.done) return
    callback(next.value)
  }
}
